### DISCUSSION TITLE
IterationAdaptiveDT multiple constraints conflict

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#23645] [Q&A General] [answered]
In `IterationAdaptiveDT`, you can use a postprocessor to constrain the time step, but it also takes time/dt pairs to force the time step at those specified times. It looks like the postprocessor takes precedence over time points, meaning if you're at a time where you specified the time step, but the postprocesser gives a smaller value, the postprocessor will win. Is there a way to force the time/dt pairs to take precedence at the given time points, or ignore the postprocessor at the given time points?

**Comment by GiudGiud** [comment]
Hello

I dont understand the concern. Could you please paste the TimeStepper block and the time steps it takes?

Right now I feel like it s obeying both constraints naturally ?

Guillaume

**Reply by jessecarterMOOSE** [reply]
Can't get to my input file from here but basically I've got:

```
  [TimeStepper]
    type = IterationAdaptiveDT
    ...
    dt = 1.0
    timestep_limiting_postprocessor = time_pps
    time_t = '1.0'
    time_dt = '20.0'
  []
```

So when I'm at t=1.0, I want to force dt = 20.0, but my time postprocessor is returning a smaller value, say 5.0.

The timestepper prints:

```
Setting dt to value specified by dt function: 20.0
Limiting dt to postprocessor value. dt = 5.0.
```

I want the time_dt value(s) to take precedence over the postprocessor. My logic is that I want to force a time step at the given time(s). Currently the time points are being ignored.

**Reply by GiudGiud** [reply]
Why even specify a timestep limiting pp if you dont want it to matter though?

The time points will be reached, it s just the time_dt that is not met. 
If you say you want to hit t = 1, 8, 12 and 20s, it will hit them even with a timestep limited to 5s by the PP.
It would do: 1, 6, 8, 12, 17, 20, taking the max PP step or the step needed to meet the desired time points

**Reply by jessecarterMOOSE** [reply]
My PP is MaterialTimeStepPostprocessor, which lags the solution time. I'm trying to force a big jump in time at t=1, but the values from my material models are based on the previous time increment, which was much smaller.  I know when the big jump times are, so I want to overrule my time PP at the specified points.

**Reply by GiudGiud** [reply]
I m not sure what the solution is here.
We are not going to change the default behavior of that Timestepper as it makes sense. 

We could add a boolean to change the priority of time stepping decisions but it's pretty custom.
You could write a custom PP that has complex enough logic to sometimes give the MatTimeStepPP value, sometimes have a larger value

**Reply by jessecarterMOOSE** [reply]
Seems like most options for that Timestepper are for _limiting_ the timestep, i.e., pick the _smallest_ value of all these constraints, but not enough options to make it _grow_. Specifying the time points was the solution [for this question](https://github.com/idaholab/moose/discussions/23145), but that strategy does work in conjunction with a postprocessor. In the case I mention above my postprocessor is actually trying to increase the timestep, but I'd like to force it to increase more, so I want the _largest_ of the constraints.

> You could write a custom PP that has complex enough logic to sometimes give the MatTimeStepPP value, sometimes have a larger value

This is what I'm thinking at this point.

**Reply by jessecarterMOOSE** [reply]
Maybe if `timestep_limiting_postprocessor` forces a max dt value, there could be an analogous postprocessor that forces a _min_ dt value? Just a thought.

**Reply by GiudGiud** [reply]
but you want to be limiting the pp with this material change? I dont think the pp you have would fit naturally for a "grower"

**Reply by jessecarterMOOSE** [reply]
I'm using an Abaqus UMAT, which returns a value called PNEWDT. If the value is <1 it is suggesting to cut the time step and if >1 it is suggesting to grow the time step. The AbaqusUMATStress class captures this value multiplied by the current dt in a timestep material, so MaterialTimeStepPostprocessor will reflect the minimum suggested time step from all the qp's. This min value may be greater than the current time step, meaning the timestep can grow.

**Reply by GiudGiud** [reply, marked as answer]
I would bake all the logic you want in a ParsedPostprocessor and use a PostprocessorDT

**Reply by jessecarterMOOSE** [reply]
I still want to use the iteration adaptive part.

**Reply by GiudGiud** [reply]
Would you prefer adding a "timestep_growth_pp" parameter or adding a modified version of IterationAdaptiveDT to your app?

**Reply by jessecarterMOOSE** [reply]
Not sure. I'd prefer the time_t/time_dt parameters to always be honored, so if I can make that happen easily with a modified IterationAdaptiveDT then that might be my preferred route. Maybe inherit from IterationAdaptiveDT and only call base class `constrainDT` if not at a time point?

**Comment by jessecarterMOOSE** [comment]
On a related note, why is `cutback_factor` [used on a failed solve](https://github.com/idaholab/moose/blob/202b5d2915a4b339731c1a5bc50e7c4303bfd3a5/framework/src/timesteppers/IterationAdaptiveDT.C#L331) instead of `cutback_factor_at_failure`? I thought `cutback_failure` was for when solves were _difficult_ but did eventually converge. 

Looking closer there's some overlap in the doc strings. Both mention failure, so it's unclear to me what the different situations are.
[cutback_factor](https://github.com/idaholab/moose/blob/202b5d2915a4b339731c1a5bc50e7c4303bfd3a5/framework/src/timesteppers/IterationAdaptiveDT.C#L67-L71): "Factor to apply to timestep if difficult convergence (if 'optimal_iterations' is specified) or if solution failed"
[cutback_factor_at_failure](https://github.com/idaholab/moose/blob/202b5d2915a4b339731c1a5bc50e7c4303bfd3a5/framework/src/timesteppers/TimeStepper.C#L21-L25): "Factor to apply to timestep if a time step fails to converge."

**Reply by GiudGiud** [reply]
oh they're some sort of overlap here. 
Thanks the report I ll get a fix

**Reply by GiudGiud** [reply]
https://github.com/idaholab/moose/pull/23660 has the fix for this.
